Introducing bindMode in BGPConfiguration

hanamantagoudvk opened the pull request:
## Description

Introducing new config parameter `bindMode` in `BGPConfiguration` to let user choose whether to use 0.0.0.0 as bind address or calico's Node.Spec.BGP.IPvXAddress.

## Related issues

https://github.com/projectcalico/calico/issues/4974

```release-note
Allow configuration of bind mode via the BGPConfiguration API
```

lwr20 commented:
@caseydavenport any thoughts on this?

neiljerram commented:
@lwr20 @hanamantagoudvk Sorry for my slowness here, I'm taking a look now.  (Of course @caseydavenport is welcome to review too!)

neiljerram commented on api/pkg/apis/projectcalico/v3/bgpconfig.go:
```diff
@@ -80,6 +80,9 @@ type BGPConfigurationSpec struct {
 	ListenPort uint16 `json:"listenPort,omitempty" validate:"omitempty,gt=0" confignamev1:"listen_port"`
+
+	// BindMode indicates whether ListenPort should bind to 0.0.0.0 or calico's Node.Spec.BGP.IPvXAddress
+	BindMode string `json:"bindMode,omitempty" validate:"omitempty,bindMode" confignamev1:"bind_mode"`
```
Also `confignamev1` isn't needed (actually, on any of these fields), because BGPConfiguration is now streamed as a v3 object.

neiljerram commented on confd/pkg/backends/calico/client.go:
```diff
@@ -959,6 +959,7 @@ func (c *client) updateBGPConfigCache(resName string, v3res *apiv3.BGPConfigurat
 	if resName == globalConfigName {
 		c.getPrefixAdvertisementsKVPair(v3res, model.GlobalBGPConfigKey{})
 		c.getListenPortKVPair(v3res, model.GlobalBGPConfigKey{}, updatePeersV1, updateReasons)
+		c.getBindAddressVPair(v3res, model.GlobalBGPConfigKey{}, updatePeersV1, updateReasons)
```
Missing "K" - does this compile?

hanamantagoudvk replied:
modified

neiljerram commented on confd/pkg/backends/calico/client.go:
```diff
@@ -1086,6 +1087,31 @@ func (c *client) getListenPortKVPair(v3res *apiv3.BGPConfiguration, key interfac
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(bindAddrV4))
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(bindAddrV6))
+	}
+	*updatePeersV1 = true
```
This is obviously missing code as it stands.  But anyway I would suggest just propagating the BindMode value so that the template can see it:
```
bindMode := getBGPConfigKey("bind_mode", key)
c.updateCache(api.UpdateTypeKVUpdated, getKVPair(bindMode, v3res.Spec.BindMode))
```

Then the templating should check for bind_mode in a similar fashion to this:
```
{{- $node_listen_port_key := printf "/host/%s/listen_port" (getenv "NODENAME")}}
{{- if exists $node_listen_port_key}}
```
If it exists and equals NodeIP, emit the `listen bgp` line with `$node_ip` as the address.

hanamantagoudvk replied:
yes modified accordingly. But when i apply below config, its still not able to recognize `bindMode` . Did i miss anything ?

```

apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  listenPort: 177
  bindMode: NodeIP
```

neiljerram replied:
@hanamantagoudvk Can you tell me exactly how you've tested, so that I can think it through?

hanamantagoudvk replied:
@neiljerram  As per Shaun's suggestion, i built `calicoctl` . With new binary also, i get the same above error.

neiljerram replied:
Ah, I think you misunderstood my comment about replacing "bindMode" with "oneof...".  I meant only for the "validate" part of the tag.  So the whole BindMode field should be:
```
	BindMode string `json:"bindMode,omitempty" validate:"omitempty,oneof=None NodeIP"`
```
It's possible there might still be other issues, but you definitely need that one corrected.

hanamantagoudvk replied:
@neiljerram @fasaxc after modifying `bgpconfig.go` as per suggestion, i rebuilt `typa `and `calicoctl`. After that i am not facing above error issue. I could see that value of bindMode is properly being set in caches. Its also being returned by `GetValues `method. But only thing is , i dont see that its reflecting in `/etc/calico/confd/config/bird.cfg` eventually.

neiljerram replied:
I'm afraid you'll need to rebuild confd as well, and the calico-node image if that is what you are using for testing.  Can you tell us more about your test procedure?

hanamantagoudvk replied:
@neiljerram , yes i am rebuilding confd and calico-node images. Then i am deploying calico-node daemonset using this  image in k8s cluster. Then as i said earlier , i apply the `bgp-cfg.yaml` (see above) which has bindMode set to NodeIP. Then when i check the `netstat -an | grep 179`  on worker nodes, i should see that bird is listening on `<node-ip>:179` . But i dont see that happening . I see `0.0.0.0:179`.

neiljerram replied:
I suggest adding some `log.Info` lines into `getBindModeKVPair`, then check the confd logging (which is part of the calico-node log) to see if your logic is really being hit and behaving correctly.

neiljerram reviewed (changes requested):
Thanks, this is a useful start - hope that my comments make sense!

hanamantagoudvk commented:
@neiljerram 

yes modified accordingly. But when i apply below config, its still not able to recognize `bindMode `. Did i miss anything ?


```
apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  listenPort: 177
  bindMode: NodeIP
```

fasaxc commented on api/pkg/apis/projectcalico/v3/bgpconfig.go:
```diff
@@ -80,6 +80,10 @@ type BGPConfigurationSpec struct {
+
+	// BindMode indicates whether to listen for BGP connections on all addresses (None)
+	// or only on the node's canonical IP address Node.Spec.BGP.IPvXAddress (NodeIP).
+	BindMode string `json:"oneof=None NodeIP"`
```
The `json` tag is wrong here, need the name of the field `"bindMode"` and you should have a validation tag too to do the oneof part.  I think both should have omitempty.

neiljerram commented on confd/etc/calico/confd/templates/bird.cfg.template:
```diff
@@ -42,13 +42,27 @@ router id {{if eq "hash" ($router_id) -}}
 {{- end}}
 
+{{- $bind_mode := ""}}
+{{- $bind_mode_key := printf "/host/%s/bind_mode" (getenv "NODENAME")}}
```
Ah, I think the problem is that the client.go code is only handling `BindMode` in the global BGPConfiguration (i.e. the BGPConfiguration with name "default"), but here the template is checking for a per-node `bind_mode` value.

`BindMode` in global BGPConfiguration -> template path "/global/bind_mode".

`BindMode` in per-node BGPConfiguration for node XYZ -> template path "/host/XYZ/bind_mode".

I think you want global, so please try updating this template coding to use "/global/bind_mode".

hanamantagoudvk replied:
I have modified the template coding to use "/global/bind_mode". Inspite of that , i see that its not being set in bird.cfg.template. I put debug logs in both `getBindModeKVPair `and `GetValues `, i see that value is being set properly. I am pasting logs from GetValues method.

We can see `/calico/bgp/v1/global/bind_mode:NodeIP`

`2022-01-21 07:15:10.515 [INFO][90] confd/client.go 1421: Returning (map[/calico/bgp/v1/global/as_num:64512 /calico/bgp/v1/global/bind_mode:NodeIP /calico/bgp/v1/global/listen_port:176 /calico/bgp/v1/global/loglevel:info /calico/bgp/v1/global/node_mesh:{"enabled": true} ...... ]) results`

neiljerram commented:
@hanamantagoudvk I had a play, please see here: https://github.com/Nordix/calico/compare/bgp-mesh...neiljerram:bgp-mesh

The main points were:
- needed a `make generate` to update some more autogenerated files.
- `$bind_mode := ` should be `$bind_mode =` in the template, because scoping behaves similarly as in Go.
- need `{{$node_ip}}` to insert the node IP in the output, instead of `$node_ip`.

I also noticed that BindMode will only work when ListenPort is also configured.  I guess it would be nice if BindMode can work without ListenPort as well - does that make sense?

FYI, I was testing by running `make test` in the `confd` subdir.

neiljerram commented:
With those commits as above, I see an expected test failure like this:
```
Populating calico with test data using calicoctl:  explicit_peering/global
Successfully applied 8 resource(s)
Starting Typha
comparing templates attempt 1
comparing templates attempt 2
comparing templates attempt 3
comparing templates attempt 4
comparing templates attempt 5
comparing templates attempt 6
comparing templates attempt 7
comparing templates attempt 8
comparing templates attempt 9
comparing templates attempt 10
Failed: bird.cfg templates do not match, showing diff of expected vs received
9a10,11
> bind_mode exists
> and is NodeIP
11c13
< listen bgp port 150; expect something here!
---
> listen bgp 10.192.0.2 port 150;
100a103
> 
```
That `listen bgp 10.192.0.2 port 150;` confirms that the mechanism is now basically working.

hanamantagoudvk commented:
> With those commits as above, I see an expected test failure like this:
> 
> ```
> Populating calico with test data using calicoctl:  explicit_peering/global
> Successfully applied 8 resource(s)
> Starting Typha
> comparing templates attempt 1
> comparing templates attempt 2
> comparing templates attempt 3
> comparing templates attempt 4
> comparing templates attempt 5
> comparing templates attempt 6
> comparing templates attempt 7
> comparing templates attempt 8
> comparing templates attempt 9
> comparing templates attempt 10
> Failed: bird.cfg templates do not match, showing diff of expected vs received
> 9a10,11
> > bind_mode exists
> > and is NodeIP
> 11c13
> < listen bgp port 150; expect something here!
> ---
> > listen bgp 10.192.0.2 port 150;
> 100a103
> > 
> ```
> 
> That `listen bgp 10.192.0.2 port 150;` confirms that the mechanism is now basically working.

When i tested with latest changes , i get following errors:

a) When i deploy calico-node daemonset, i see syntax errors like below
```
bird: /etc/calico/confd/config/bird.cfg:10:19 syntax error
bird: /etc/calico/confd/config/bird6.cfg:10:19 syntax error
```

b) When i ran `make test` in `confd `directory , with debug logs added in template file, i see the latest logic not working. Is there any problem in latest template logic ? I modified the template because checking `NodeIP `was entrenched inside `listen_port `logic. Now both are independent.

neiljerram commented:
@hanamantagoudvk Re (a): yes, some of my edits in the template files were only for experimentation, and will definitely cause syntax errors if they reach BIRD!  Please review my commits carefully to make sure you understand them, and then only apply the changes that are strictly needed.

hanamantagoudvk commented:
> @hanamantagoudvk Re (a): yes, some of my edits in the template files were only for experimentation, and will definitely cause syntax errors if they reach BIRD! Please review my commits carefully to make sure you understand them, and then only apply the changes that are strictly needed.

Yes agreed @neiljerram . I have not put debug messages in template file while compiling image for calico-node daemonset. Is there a tool to verify the template file syntax ?

hanamantagoudvk commented:
@neiljerram , i am finding it difficult to understand whats wrong in template files. Could you please point to any material  for this template language?

hanamantagoudvk commented:
@neiljerram , this template change is not working after making `listen_port `and `bind_mode `independent. Could you please let me know whats wrong here ? I am not able to figure out the issue in template

neiljerram commented:
@hanamantagoudvk Please see https://github.com/neiljerram/calico/commit/af67ba72a63ebc9905eb0a50183666741292330f, and a sample test at https://github.com/neiljerram/calico/commit/763eff04d5ffe1954c250b582b54a715fe8d058e.  Documentation for the template language is here: https://pkg.go.dev/text/template.

Please can I also ask: can you only merge or add commits from now on, instead of squashing and/or rebasing?  It makes it easier for me to see what you have changed since our previous discussion point.

hanamantagoudvk commented:
> @hanamantagoudvk Please see [neiljerram@af67ba7](https://github.com/neiljerram/calico/commit/af67ba72a63ebc9905eb0a50183666741292330f), and a sample test at [neiljerram@763eff0](https://github.com/neiljerram/calico/commit/763eff04d5ffe1954c250b582b54a715fe8d058e). Documentation for the template language is here: https://pkg.go.dev/text/template.
> 
> Please can I also ask: can you only merge or add commits from now on, instead of squashing and/or rebasing? It makes it easier for me to see what you have changed since our previous discussion point.

@neiljerram : I have incorporated the template changes you have done. Its working partially. Here are the below steps of testing and let me know whether the steps and expectations are correct.

a) Did fresh deployment of k8s cluster with calico.
b) Checked no custom resource is configured for `bgpconfigurations.crd.projectcalico.org` 

```
# kubectl get bgpconfigurations.crd.projectcalico.org -A
No resources found
```

c)  Configure CR (default) with listenPort:176 and bindMode: NodeIP
```
apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  listenPort: 176
  bindMode: NodeIP
```

d) Checked on worker nodes. I could see `listen_port `is changed to 176 and its using `NodeIP `to bind i.e `NodeIP:176`

e) But when i modify CR with bindMode:None and apply , i see that it doesnt fallback to `0.0.0.0:176`

With these code changes and template changes , we can do dynamic updates right ? It is supposed to work right ?

neiljerram commented:
@hanamantagoudvk Yes, it definitely should be dynamic, and so should update back to 0.0.0.0:176.

neiljerram commented:
How are you seeing `NodeIP:176`?  I guess with `ss` or `netstat`.

When you switch back to `bindMode: None`, I suggest some checks:
- Read back the CR to check it says `bindMode: None`
- Look at the generated /etc/calico/confd/config/bird.cfg on one of the Calico nodes (inside the calico-node pod) to check its `listen bgp` line.
- Look at the calico-node log to see if there are any clues there about confd handling the change back from `NodeIP` to `None`.

hanamantagoudvk commented:
> How are you seeing `NodeIP:176`? I guess with `ss` or `netstat`.
>

@neiljerram , i am checking using `netstat`
 
> When you switch back to `bindMode: None`, I suggest some checks:
> 
> * Read back the CR to check it says `bindMode: None`

Yes , when i check CR , i see that `bindMode:None`

> * Look at the generated /etc/calico/confd/config/bird.cfg on one of the Calico nodes (inside the calico-node pod) to check its `listen bgp` line.

Generated `/etc/calico/confd/config/bird.cfg` is having `listen bgp port 176; `

> * Look at the calico-node log to see if there are any clues there about confd handling the change back from `NodeIP` to `None`.

Below logs from `calico-node` when i modify `bindMode: NodeIP` to `bindMode:None`

Important log to notice is `Reconfiguration of BGP listening socket not implemented, please restart BIRD.`

```
2022-02-03 05:35:49.912 [INFO][89] confd/client.go 902: Recompute BGP peerings: listenPort updated.; bindMode updated.; AS number deleted.
...
...
2022-02-03 05:35:49.914 [INFO][84] felix/int_dataplane.go 1520: Received *proto.GlobalBGPConfigUpdate update from calculation graph msg=
bird: Reconfiguration requested by SIGHUP
bird: Reconfiguring
bird: Reconfiguration of BGP listening socket not implemented, please restart BIRD.
bird: device1: Reconfigured
bird: direct1: Reconfigured
bird: Mesh_10_4_0_21: Reconfigured
bird: Mesh_10_4_0_61: Reconfigured
bird: Reconfigured
2022-02-03 05:35:49.929 [INFO][89] confd/resource.go 278: Target config /etc/calico/confd/config/bird.cfg has been updated due to change in key: /calico/bgp/v1/global
bird: Reconfiguration requested by SIGHUP
bird: Reconfiguring
bird: Reconfiguration of BGP listening socket not implemented, please restart BIRD.
bird: device1: Reconfigured
bird: direct1: Reconfigured
bird: Mesh_2001_db8_100_a1__24d: Reconfigured
bird: Mesh_2001_db8_100_a1__36e: Reconfigured
bird: Reconfigured
2022-02-03 05:35:49.932 [INFO][89] confd/resource.go 278: Target config /etc/calico/confd/config/bird6.cfg has been updated due to change in key: /calico/bgp/v1/global
```

hanamantagoudvk commented:
@neiljerram : In continuation with testing ( steps a,b,c,d,e as explained above) , after step-(e) , the expectation was to see `0.0.0.0:176`, but that doesn't happen . But after this when i update the default CR with port 175, i see that `listen_address `gets changed to `0.0.0.0:175`. So the enforcement of `bindMode: None` happens when port get updated. We want to make each of them (listen_address and listen_port) independent , but its not happening.

neiljerram commented:
@hanamantagoudvk This is the key point, which I'm afraid I did not know about before:
> bird: Reconfiguration of BGP listening socket not implemented, please restart BIRD.

Reviewing the BIRD code, I can see that:
- There is a global `bgp_listen_sk`, which is initially NULL.
- `bgp_listen_sk` is opened when the first `bgp protocol` instance is configured, according to the `listen bgp` configuration at that time.
- `bgp_listen_sk` is closed and reset to NULL if all `bgp protocol` instances are removed from the config.

Your note
> Generated /etc/calico/confd/config/bird.cfg is having listen bgp port 176;

confirms that the confd code is correctly processing updates as far as updating the bird.cfg.

But, given the BIRD code notes above, an updated `listen bgp` configuration can only take effect if all `bgp protocol` config is removed and then later re-added.  Or if the whole BIRD daemon is stopped and restarted.  I wonder if that is consistent with what was happening when you saw 0.0.0.0:175.

@caseydavenport Did you know about this?  (Is there perhaps some support that I've missed for restarting BIRD in some cases?)

caseydavenport commented:
I did not know about that, but it certainly complicates things. Normally users expect the API change to take effect without restarting the components, and we can't really restart BIRD under the covers without potentially sacrificing a graceful restart (e.g,. if the config change impacts multiple peers at once)

neiljerram commented:
Thanks @caseydavenport .  Right now there is an inconsistency in my understanding: @hanamantagoudvk has reported that it works to change the listen port setting, but I don't see from the code how that can work.  So I must still be missing something here; will try to find time next week to investigate further.

hanamantagoudvk commented:
@neiljerram , were you able to figure out something ?

neiljerram commented:
@hanamantagoudvk I'm sorry, no.  I wonder if it might be easier to enhance the BIRD code to handle this case, than to work around the deficiency in confd code.  If you'd like to look into that, our fork of the BIRD code is at https://github.com/projectcalico/bird.

hanamantagoudvk commented:
> @hanamantagoudvk I'm sorry, no. I wonder if it might be easier to enhance the BIRD code to handle this case, than to work around the deficiency in confd code. If you'd like to look into that, our fork of the BIRD code is at https://github.com/projectcalico/bird.

@neiljerram , let me know how do i proceed on this ? i am totally stuck and blank on this one.

neiljerram commented:
@hanamantagoudvk Important caveat: I don't yet know for sure if this will be a reasonable change to work on.  But if I was doing this, I would start by reviewing the code related to:
1. The BGP listen socket - look for `bgp_listen_sk`.  (I also talked about this a bit 6 comments above this one.)
2. What happens when BIRD processes a new configuration - look for `Reconfiguration of BGP listening socket not implemented` to find a way in to that code.

Hope that's helpful!

hanamantagoudvk commented:
@neiljerram @caseydavenport : I analyzed  BIRD code further along with logs i collected with various changes. So i see some interesting behavior. So i will explain here.

a) After initial bring up of k8s cluster with calico , BGP stack listens on 0.0.0.0:179.

b) Applied below CR . 

```
# cat bgp-cfg.yaml
apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  listenPort: 174
  bindMode: NodeIP
```

From the BIRD logs below , i see that there is a state transition happening from `STOP `--> `DOWN `--> `START`.
I assume this is restarting connections.

```
bird: Reconfiguration requested by SIGHUP
bird: Reconfiguring
bird: Reconfiguration of BGP listening socket not implemented, please restart BIRD.
bird: device1: Reconfigured
bird: direct1: Reconfigured
bird: Restarting protocol Mesh_2001_db8_100_a1__24d
bird: Mesh_2001_db8_100_a1__24d: Shutting down                 <-----------
bird: Mesh_2001_db8_100_a1__24d: State changed to stop     <-----------
bird: Restarting protocol Mesh_2001_db8_100_a1__36e
bird: Mesh_2001_db8_100_a1__36e: Shutting down
bird: Mesh_2001_db8_100_a1__36e: State changed to stop     <-----------
bird: Mesh_2001_db8_100_a1__24d: State changed to down   <----------
bird: Mesh_2001_db8_100_a1__24d: Initializing
bird: Mesh_2001_db8_100_a1__24d: Starting
bird: Mesh_2001_db8_100_a1__24d: State changed to start     <-----------
bird: Mesh_2001_db8_100_a1__36e: State changed to down   <-----------
bird: Mesh_2001_db8_100_a1__36e: Initializing
bird: Mesh_2001_db8_100_a1__36e: Starting
bird: Mesh_2001_db8_100_a1__36e: State changed to start    <------------
bird: Reconfigured
bird: Reconfiguration requested by SIGHUP
bird: Reconfiguring
bird: Reconfiguration of BGP listening socket not implemented, please restart BIRD.
bird: device1: Reconfigured
2022-03-10 07:59:17.570 [INFO][93] confd/resource.go 278: Target config /etc/calico/confd/config/bird.cfg has been updated due to change in key: /calico/bgp/v1/global
bird: direct1: Reconfigured
bird: Restarting protocol Mesh_10_4_0_21
bird: Mesh_10_4_0_21: Shutting down
bird: Mesh_10_4_0_21: State changed to stop
bird: Restarting protocol Mesh_10_4_0_61
bird: Mesh_10_4_0_61: Shutting down
Mesh_10_4_0_61: State changed to stop
2022-03-10 07:59:17.570 [INFO][93] confd/resource.go 278: Target config /etc/calico/confd/config/bird6.cfg has been updated due to change in key: /calico/bgp/v1/global
bird: Mesh_10_4_0_21: State changed to down

bird: Mesh_10_4_0_21: Starting
bird: Mesh_10_4_0_21: State changed to start       <------------
bird: Mesh_10_4_0_61: State changed to down     <------------
bird: Mesh_10_4_0_61: Initializing
bird: Mesh_10_4_0_61: Starting
bird: Mesh_10_4_0_61: State changed to start      <------------
bird: Reconfigured
```

c) Now modified the CR to `bindMode: None` and keeping the port same at 174


```
# cat bgp-cfg.yaml
apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  listenPort: 174
  bindMode: None
```

From the below BIRD logs, i dont see any state transition happening like in step-(b)

```
bird: Reconfiguration requested by SIGHUP
bird: Reconfiguring
bird: Reconfiguration of BGP listening socket not implemented, please restart BIRD.
bird: device1: Reconfigured
2022-03-10 08:21:00.994 [INFO][93] confd/resource.go 278: Target config /etc/calico/confd/config/bird.cfg has been updated due to change in key: /calico/bgp/v1/global
bird: direct1: Reconfigured
bird: Mesh_10_4_0_21: Reconfigured
bird: Mesh_10_4_0_61: Reconfigured
bird: Reconfigured
bird: Reconfiguration requested by SIGHUP
2022-03-10 08:21:00.996 [INFO][93] confd/resource.go 278: Target config /etc/calico/confd/config/bird6.cfg has been updated due to change in key: /calico/bgp/v1/global
bird: Reconfiguring
bird: Reconfiguration of BGP listening socket not implemented, please restart BIRD.
bird: device1: Reconfigured
bird: direct1: Reconfigured
bird: Mesh_2001_db8_100_a1__24d: Reconfigured
bird: Mesh_2001_db8_100_a1__36e: Reconfigured
bird: Reconfigured
```

d) 	Now modified the CR keeping bindMode: None and port value changed to 173

```
apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  listenPort: 173
  bindMode: None
```

From below BIRD logs, i see that state transition happens from `STOP `--> `DOWN `--> `START`.
From these observations , i can say that whenever there is proper state transition , config changes done is getting
reflected properly. Any thoughts on this @neiljerram @caseydavenport 


```
bird: Mesh_10_4_0_61: Received: Other configuration change
bird: Mesh_10_4_0_61: State changed to stop                     <----------------
bird: Mesh_10_4_0_61: State changed to down                   <-----------------
bird: Mesh_10_4_0_61: Starting
bird: Mesh_10_4_0_61: State changed to start
bird: Reconfiguration requested by SIGHUP
bird: Reconfiguration requested by SIGHUP
bird: Reconfiguring
bird: Reconfiguration of BGP listening socket not implemented, please restart BIRD.
bird: device1: Reconfigured
bird: direct1: Reconfigured
bird: Restarting protocol Mesh_2001_db8_100_a1__24d
bird: Mesh_2001_db8_100_a1__24d: Shutting down
bird: Mesh_2001_db8_100_a1__24d: State changed to stop
bird: Restarting protocol Mesh_2001_db8_100_a1__36e
bird: Mesh_2001_db8_100_a1__36e: Shutting down
bird: Mesh_2001_db8_100_a1__36e: State changed to stop      <----------------
bird: Mesh_2001_db8_100_a1__24d: State changed to down    <---------------
bird: Mesh_2001_db8_100_a1__24d: Initializing
bird: Mesh_2001_db8_100_a1__24d: Starting
bird: Mesh_2001_db8_100_a1__24d: State changed to start     <----------------
bird: Mesh_2001_db8_100_a1__36e: State changed to down    <---------------
bird: Mesh_2001_db8_100_a1__36e: Initializing
bird: Mesh_2001_db8_100_a1__36e: Starting
bird: Mesh_2001_db8_100_a1__36e: State changed to start
bird: Reconfigured
bird: Reconfiguring
bird: Reconfiguration of BGP listening socket not implemented, please restart BIRD.
bird: device1: Reconfigured
bird: direct1: Reconfigured
bird: Restarting protocol Mesh_10_4_0_21
bird: Mesh_10_4_0_21: Shutting down
bird: Mesh_10_4_0_21: State changed to stop
bird: Restarting protocol Mesh_10_4_0_61
bird: Mesh_10_4_0_61: Shutting down
bird: Mesh_10_4_0_61: State changed to stop
2022-03-10 14:16:05.159 [INFO][93] confd/resource.go 278: Target config /etc/calico/confd/config/bird6.cfg has been updated due to change in key: /calico/bgp/v1/global
bird: Mesh_10_4_0_61: State changed to down
bird: Mesh_10_4_0_61: Initializing
bird: Mesh_10_4_0_61: Starting
bird: Mesh_10_4_0_61: State changed to start                    <----------------
bird: Mesh_10_4_0_21: State changed to down                  <----------------
bird: Mesh_10_4_0_21: Initializing                                      
bird: Mesh_10_4_0_21: Starting
bird: Mesh_10_4_0_21: State changed to start                     <-----------
bird: Reconfigured
```

hanamantagoudvk commented:
@neiljerram @caseydavenport : please give your suggestion based on above observation. If you think , modifying BIRD code is not an easy or preferred option , is it possible that we go with on below mentioned option ?

a) confd code changes to be merged.
b) Document needs to be updated where restarting of BIRD manually by admin needs to be recommended.

caseydavenport commented:
It's not ideal, but we might be able to justify restarting BIRD for this because:
- It's a niche use case and should mostly be set at cluster creation, not throughout a cluster's life. 
- We can probably get the operator to roll calico/node gracefully, so at least operator-managed clusters can automatically do this.

We would probably need the operator to monitor BGPConfiguration and include an annotation to trigger rolling update when this particular field changed.

hanamantagoudvk commented:
> 


> It's not ideal, but we might be able to justify restarting BIRD for this because:
> 
> * It's a niche use case and should mostly be set at cluster creation, not throughout a cluster's life.
> * We can probably get the operator to roll calico/node gracefully, so at least operator-managed clusters can automatically do this.
> 
> We would probably need the operator to monitor BGPConfiguration and include an annotation to trigger rolling update when this particular field changed.

hanamantagoudvk commented:
> It's not ideal, but we might be able to justify restarting BIRD for this because:
> 
> * It's a niche use case and should mostly be set at cluster creation, not throughout a cluster's life.
> * We can probably get the operator to roll calico/node gracefully, so at least operator-managed clusters can automatically do this.
> 
> We would probably need the operator to monitor BGPConfiguration and include an annotation to trigger rolling update when this particular field changed.


@caseydavenport : If i understand you correctly:

a)  For operator based calico installation (i.e Tigera operator) , the same operator needs to be enhanced to watch for BGP config and restart calico/node.

b) For non-operator based calico installation , the BGP config needs to be set at the cluster creation time.
No further modifications supported during the lifetime of cluster. 

Is this correct ?

caseydavenport commented:
Yep, that is what I was trying to say with one slight change:

> No further modifications supported during the lifetime of cluster.

Modifications could be supported so long as the user triggers a rolling update manually.

hanamantagoudvk commented:
Yes agree with you here.

Regarding enhancements in operator based approach , i will raise PR for it ? Any pointers for it ?

hanamantagoudvk commented:
@caseydavenport : I am trying to make code changes in Tigera operator code. Have few questions.

a) I have to watch for `BGPConfiguration` resource. Whenever there is a reconcile , i need to restart the calico-node .. right ?
b) Or Do i have use sharedInformer to watch the resource and only if there is update to the resource , i need to restart ?

caseydavenport commented:
/sem-approve

caseydavenport commented on api/pkg/apis/projectcalico/v3/bgpconfig.go:
```diff
@@ -91,6 +91,10 @@ type BGPConfigurationSpec struct {
+
+	// BindMode indicates whether to listen for BGP connections on all addresses (None)
+	// or only on the node's canonical IP address Node.Spec.BGP.IPvXAddress (NodeIP).
+	BindMode string `json:"bindMode,omitempty" validate:"omitempty,oneof=None NodeIP"`
```
For API fields like this, we should be defining a new type. e.g.,

```go
type BindMode string

const (
    BindModeNone BindMode = "None"
    BindModeNodeIP BindMOde = "NodeIP"
)
```

Since this is an optional field, per Kubernetes API conventions, this should also be a pointer and be marked as optional.

e.g.,

```go
// +optional
BindMode *BindMode `json:"bindMode,omitempty"`
```

caseydavenport replied:
The comment should also include the default behavior for the reader's benefit.

neiljerram replied:
I'm afraid I gave @hanamantagoudvk the opposite feedback, earlier on this PR.  @caseydavenport Are we sure that the additional boilerplate for **not** using `oneof=` is really useful?

caseydavenport replied:
I think stylistically it can come down to preference, but this comes straight from the Kubernetes API conventions doc which we try to follow best we can: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types

caseydavenport reviewed (changes requested):
@hanamantagoudvk this is looking pretty close to me. I added a couple of comments on the API section that need to be addressed in order for the API to be consistent with our conventions. 

We also need a new test added in `confd/tests` for this scenario.

@neiljerram did you have any other feedback? Otherwise I think we're good to approve as soon as the API changes are in, and confd tests are updated.

caseydavenport commented:
v3.23 code freeze is this Friday - would be nice to get it in before then but otherwise we will have do this for v3.24. Let me know if I can help get it over the line.

neiljerram reviewed (commented):
Two thoughts, but generally this looks good to me.  Thanks @hanamantagoudvk for persisting with it!

hanamantagoudvk commented on calico/reference/resources/bgpconfig.md:
```diff
@@ -56,6 +57,7 @@ spec:
 | serviceExternalIPs | The CIDR blocks for Kubernetes Service External IPs to be advertised over BGP. Kubernetes Service External IPs will only be advertised if they are within one of these blocks. Only valid on the global `default` BGPConfiguration: will be ignored otherwise. | A list of valid IPv4 or IPv6 CIDR blocks. | List of `cidr: <ip>/<prefix length>` values. | Empty List |
 | serviceLoadBalancerIPs | The CIDR blocks for Kubernetes Service status.LoadBalancer IPs to be advertised over BGP. Kubernetes LoadBalancer IPs will only be advertised if they are within one of these blocks. Only valid on the global `default` BGPConfiguration: will be ignored otherwise. | A list of valid IPv4 or IPv6 CIDR blocks. | List of `cidr: <ip>/<prefix length>` values. | Empty List |
 | listenPort | The port where BGP protocol should listen.| A valid port number. | integer | 179 |
+| bindMode | Indicates whether to listen for BGP connections on all addresses (None) or only on the node's canonical IP address Node.Spec.BGP.IPvXAddress (NodeIP).| None, NodeIP. | string | None |
```
@neiljerram  Yes we need to capture it somewhere. But i dont know which is the right place. Please tell me.

neiljerram replied:
I suggest adding it into that same cell: perhaps "If this field is changed when calico-node is already running, the change will not take effect until calico-node is manually restarted."  WDYT?

neiljerram replied:
I suggest adding it into that same cell: perhaps "If this field is changed when calico-node is already running, the change will not take effect until calico-node is manually restarted."  WDYT?

hanamantagoudvk replied:
@neiljerram , so its applicable only for deployment where Tigera operator is not used. So should it be very specific then ?

caseydavenport commented:
/sem-approve

hanamantagoudvk commented:
@caseydavenport @neiljerram : I am not sure where is the CI failure. I checked node related link , but not able to figure out what is the exact error .

neiljerram commented:
:tada:
